jukebox/src/lib.rs

```diff
@@ -122,31 +138,75 @@ impl Jukebox {
+                        if let Ok(entries) = read_dir(&stage_dir) {
+                            // Get all files in folder
+                            let files: Vec<_> = entries
+                                .filter_map(|entry| {
+                                    if let Ok(entry) = entry {
+                                        let path = entry.path();
+                                        if path.is_file() {
+                                            if let Some(extension) = path.extension().and_then(|extension| extension.to_str()) {
+                                                if ["mp3", "wav", "ogg", "flac"].contains(&extension.to_lowercase().as_str()) {
+                                                    return Some(path);
+                                                }
+                                            }
+                                        }
+                                    }
+                                    None
+                                })
+                                .collect();
 
-                    // Decode the Hps into audio
-                    let audio = match hps.decode() {
-                        Ok(audio) => audio,
-                        Err(e) => {
-                            tracing::error!(target: Log::Jukebox, error = ?e, "Failed to decode hps into audio. Cannot play song.");
-                            Dolphin::add_osd_message(
-                                Color::Red,
-                                OSDDuration::Normal,
-                                "Invalid music data found in ISO. This music will not play.",
-                            );
-                            continue;
-                        },
-                    };
+                            // Choose a random file from the stage folder if available
+                            if !files.is_empty() {
+                                if let Some(random_file) = files.choose(&mut rand::thread_rng()) {
+                                    custom_song_path = Some(random_file.clone())
+                                }
+                            }
+                        }
+                    }
+
+                    // Append stage audio to sink
+                    if let Some(custom_song_path) = custom_song_path {
+                        match File::open(custom_song_path) {
+                            Ok(custom_song_file) => {
+                                if let Ok(custom_song) = rodio::Decoder::new(BufReader::new(custom_song_file)) {
+                                    sink.append(custom_song.repeat_infinite());
+                                }
+                            },
+                            Err(e) => {
+                                tracing::error!(target: Log::Jukebox, error = ?e, "Failed to open custom song. Cannot play song.");
+                                continue;
+                            },
+                        }
```

Comment:
A few notes about this:
- I strongly suggest moving this logic into it's own function to keep `start()` terse and easy to read.
- In a function you can greatly improve readability by reducing nesting. Instead of:
  ```rs
  if let Some(foo) = get_foo() {
  ```
  you can simply do:
  ```rs
  let foo = get_foo()?;
  ```
- You should call as much of this as possible before the main loop as to avoid reading from the disk unnecessarily every time a new song loads.